### PR TITLE
[do not merge] docs(adr): pause unbonding peruid during equivocation proposal

### PR DESCRIPTION
**Pull request by tbruyelle**

### CODE REVIEW

File: docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+conflicts with the *Completion of Unbonding Operations* system. Giving an
+unbonding period (already paused by definition because of the *Completion of
+Unbonding Operations* system), an additional pause will just increase the
+counter, so when this unbonding period has reached its maturity on all provider
+and consumer chain, it will remain paused.
+
+### When pause
+
+The unbonding period (if there's any) should paused once an equivocation
+proposal is activated (in other words, when it enters the voting period). For
```
Comment (giunatale) [inline docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md:44]:
```suggestion
proposal enters the voting period
```

Seems more clear to me this way

File: docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+that, the `gov` module hooks `AfterProposalDeposit` can be used. 
+
+If the proposal is an equivocation proposal in voting period for the first time
```
Comment (giunatale) [inline docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md:47]: Clarify this part with more context.

File: docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+(this is important because a proposal can receive additional deposit, even
+after its activation), then for each equivocation of the proposal, the related
+unbonding operations of the related validator must be paused.
+
+### When unpause
+
+Similarly, a `gov` module hook can be used, here it's
```
Comment (giunatale) [inline docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md:54]:
```suggestion
We can use a `gov` module's hook also here and it is
```

File: docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+unbonding period (already paused by definition because of the *Completion of
+Unbonding Operations* system), an additional pause will just increase the
+counter, so when this unbonding period has reached its maturity on all provider
+and consumer chain, it will remain paused.
+
+### When pause
+
+The unbonding period (if there's any) should paused once an equivocation
+proposal is activated (in other words, when it enters the voting period). For
+that, the `gov` module hooks `AfterProposalDeposit` can be used.
```
Comment (giunatale) [inline docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md:45]:
```suggestion
that, the `gov` module's hook `AfterProposalDeposit` can be used. 
```

File: docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+`AfterProposalVotingPeriodEnded`.
+
+If the proposal is an equivocation proposal, then for each equivocation,
+unpause all unbonding operations of the related validator.
+
+## Consequences
+
+### Positive
+
+- Validators threatened by an equivocation proposal cannot play with the time to
```
Comment (giunatale) [inline docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md:64]: The meaning it's clear but I would rephrase slightly, `cannot play with the time to..` in particular should be rewritten in my opinion

File: docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  unbond their tokens before the end of the voting period.
+
+### Negative
+
+- Malicious consumer chain could forge slash packets that allow to submit an
+  equivocation proposal on the provider chain, that will freeze validor's
+  unbonding periods for an undeterminated amount of time.
+- Misbehavior on consumer can potentially being unpunished, if noone submits an
```
Comment (giunatale) [inline docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md:72]:
```suggestion
- Misbehavior on consumer chain can potentially go unpunished, if no one submits an
```

File: docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+  equivocation proposal or if the proposal doesn't pass.
```
Comment (giunatale) [inline docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md:73]:
```suggestion
  equivocation proposal in time, or if the proposal doesn't pass.
```

File: docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+## Consequences
+
+### Positive
+
+- Validators threatened by an equivocation proposal cannot play with the time to
+  unbond their tokens before the end of the voting period.
+
+### Negative
+
+- Malicious consumer chain could forge slash packets that allow to submit an
```
Comment (giunatale) [inline docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md:69]:
```suggestion
- Malicious consumer chain could forge slash packets enabling submission of an
```

File: docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+  equivocation proposal on the provider chain, that will freeze validor's
```
Comment (giunatale) [inline docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md:70]:
```suggestion
  equivocation proposal on the provider chain, resulting in the freezing of validator's
```

File: docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md (hunk 2 above)
Comment (giunatale) [reply docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md:47]: I would rewrite it to something like:
"even after the voting period has started, a proposal can receive additional deposits. The hook is triggered however at arrival of a deposit, so a check to verify that the proposal is not already in voting period is required."

File: docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md (hunk 5 above)
Comment (tbruyelle) [reply docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md]: rephrased thx!

File: docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md (hunk 2 above)
Comment (tbruyelle) [reply docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md]: thx for the suggestion, I splited into 2 sentences

File: docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+## Changelog
+* 2023-05-16: Initial Draft
+
+## Status
+
+Proposed
+
+## Context
+
+(Copied from #747)
```
Comment (giunatale) [inline docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md:16]:
```suggestion
(Copied from [#747](https://github.com/cosmos/interchain-security/issues/747))
```

File: docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+period is 3 weeks, and the voting period is 3 weeks. For this reason, it might
+be good to pause unbondings for validators named in an equivocation slashing
+proposal until the proposal's voting period is over.
+
+## Decision
+
+### How
+
+Pausing the unbonding period is already possible thanks to the changes in the
+staking module of the cosmos-sdk:
```
Comment (giunatale) [inline docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md:30]:
```suggestion
`staking` module of the cosmos-sdk:
```

File: docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+- `stakingKeeper.PutUnbondingOnHold` pauses an unbonding period
+- `stakingKeeper.UnbondingCanComplete` unpauses an unbonding period
+
+Those methods use a counter under the hood, which means we can use them without
```
Comment (giunatale) [inline docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md:34]:
```suggestion
These methods use a counter under the hood, which means we can use them without 
```

File: docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+Pausing the unbonding period is already possible thanks to the changes in the
+staking module of the cosmos-sdk:
+- `stakingKeeper.PutUnbondingOnHold` pauses an unbonding period
+- `stakingKeeper.UnbondingCanComplete` unpauses an unbonding period
+
+Those methods use a counter under the hood, which means we can use them without 
+conflicts with the *Completion of Unbonding Operations* system. Giving an
+unbonding period (already paused by definition because of the *Completion of
+Unbonding Operations* system), an additional pause will just increase the
+counter, so when this unbonding period has reached its maturity on all provider
```
Comment (giunatale) [inline docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md:38]:
```suggestion
counter, so when this unbonding period has reached its maturity on provider
```

File: docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+and consumer chain, it will remain paused.
```
Comment (giunatale) [inline docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md:39]:
```suggestion
and all consumer chains, it will remain paused.
```

File: docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+
+### When pause
+
+The unbonding period (if there's any) should be paused once an equivocation
```
Comment (giunatale) [inline docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md:43]:
```suggestion
The unbonding period (if there is any existing unboding) should be paused once an equivocation
```

File: docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+additional deposits. The hook is triggered however at arrival of a deposit, so
+a check to verify that the proposal is not already in voting period is
+required.
+
+### When unpause
+
+We can use a `gov` module's hook also here and it is
+`AfterProposalVotingPeriodEnded`.
+
+If the hook is triggered with a proposal that is an equivocation proposal, then
```
Comment (giunatale) [inline docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md:61]:
```suggestion
If the hook is triggered with an equivocation proposal, then
```

File: docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+for each equivocation, unpause all unbonding operations of the related
+validator.
+
+## Consequences
+
+### Positive
+
+- Validators threatened by an equivocation proposal cannot fnish unbonding
```
Comment (giunatale) [inline docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md:69]:
```suggestion
- Validators subject to an equivocation proposal cannot finish unbonding
```

File: docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+  their tokens before the end of the voting period.
+
+### Negative
+
+- Malicious consumer chain could forge slash packets enabling submission of an
```
Comment (giunatale) [inline docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md:74]:
```suggestion
- A malicious consumer chain could forge slash packets enabling submission of an
```

File: docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+  equivocation proposal on the provider chain, resulting in the freezing of
+  validator's unbonding periods for an undeterminated amount of time.
```
Comment (giunatale) [inline docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md:76]:
```suggestion
  validator's unbondings for an undeterminated amount of time.
```

File: docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+- Misbehavior on consumer chain can potentially go unpunished, if no one
```
Comment (giunatale) [inline docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md:77]:
```suggestion
- Misbehavior on a consumer chain can potentially go unpunished, if no one
```

File: docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md (hunk 15 above)
Comment (tbruyelle) [reply docs/docs/adrs/adr-007-pause-unbonding-on-eq-prop.md]: fixed manually bc of typo in `unboding`

**Review by giunatale — APPROVED** [review]
Looks good!

**Comment by tbruyelle** [comment]
Closed in favor of PR targeting the main repo.
